### PR TITLE
Cwillner/feat--add-suggestions-from-pnonn

### PR DESCRIPTION
**Pull request by vaeng**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -23,7 +23,9 @@ CMakeLists.txt.user
 # Project specific 'ignore patterns'
 #
 # Put you project specific ignores below.
-# If they are generally useful consider adding them to the
```
Comment (mhier) [inline .gitignore:26]: This line must be identical to the corresponding line in the project-template...

File: .gitignore [hunk 2]
```diff
@@ -23,7 +23,9 @@ CMakeLists.txt.user
...
 # Put you project specific ignores below.
-# If they are generally useful consider adding them to the 
+# If they are generally useful consider adding them to the
 # project-template and the merge them in.
 ######################
-*.code-workspace
\ No newline at end of file
+*.code-workspace
+#used by PyCharm IDE
+.idea/
```
Comment (mhier) [inline .gitignore:31]: please add missing newline at end of file (was missing before, but still...)

File: deviceaccess.py [hunk 3]
```diff
@@ -15,15 +15,16 @@
...
 from __future__ import annotations
 
-from typing import Sequence
+from typing import Sequence, Union
 import _da_python_bindings as pb
 import numpy as np
 from _da_python_bindings import AccessMode, DataValidity, TransferElementID, VersionNumber
 from abc import ABC
-
+from functools import reduce
```
Comment (mhier) [inline deviceaccess.py:23]: Python question: If I import the deviceaccess module, will I get `reduce` imported from functools as well automatically?

If yes: We should avoid all `from X import Y` or `import X as Y` constructions, as they will spoil the top-level namespace of the using Python scripts.

File: deviceaccess.py [hunk 4]
```diff
@@ -1049,7 +1088,7 @@ def getTwoDRegisterAccessor(
         Parameters
         ----------
         userType : type or numpy.dtype
-          The userType for the accessor. Can be float, or any of the numpy.dtype
+          The userType for the accessor. Can be of type float, or any of the numpy.dtype
```
Comment (mhier) [inline deviceaccess.py:1091]: no, the plain python type float is no longer possible, you replaced it with np.single...

File: deviceaccess.py [hunk 5]
```diff
@@ -1381,3 +1426,13 @@ def write(
...
+
+#######################################################################################################################
+# Type Definitions
+
+
+UserType = reduce(lambda x, y: Union[x, y], list(Device._userTypeExtensions.values()))
+
+RegisterAccessor = Union[OneDRegisterAccessor,
+                         TwoDRegisterAccessor,
+                         ScalarRegisterAccessor]
```
Comment (mhier) [inline deviceaccess.py:1438]: Shouldn't these be placed above their usage?

**Review by mhier — COMMENTED** [review]
Some minor corrections, just go ahead with merging once they are corrected.

File: deviceaccess.py (hunk 4 above)
Comment (vaeng) [reply deviceaccess.py:1091]: Good catch

File: deviceaccess.py (hunk 5 above)
Comment (vaeng) [reply deviceaccess.py:1438]: Device is not known at the start of the file, so it has to be below Device.

Annotations can be defined after they are used, as they are evaluated later. https://docs.python.org/3/howto/annotations.html

Comment (mhier) [reply deviceaccess.py:1438]: all right, thanks for the explanation.

File: deviceaccess.py (hunk 3 above)
Comment (vaeng) [reply deviceaccess.py:23]: We do want to have `AccessMode` etc. available in the imported namespace, though.
I am not sure about Sequence and Union, af this will get a bit wordy for the type hints.

I think it is pretty a pretty "normal" behaviour to have a "poluted namespace", but I agree to keep it low.
